src/wayback/_http.py

```diff
@@ -0,0 +1,591 @@
+        return super().request(*args, **kwargs)
+
+    def should_retry(self, response: WaybackHttpResponse):
+        # A memento may actually be a capture of an error, so don't retry it :P
+        if response.is_memento:
+            return False
+
+        return response.status_code in self.retryable_statuses
+
+    def should_retry_error(self, error):
+        if isinstance(error, self.retryable_errors):
+            return True
+        elif isinstance(error, ConnectionError):
+            # ConnectionErrors from requests actually wrap a whole family of
+            # more detailed errors from urllib3, so we need to do some string
+            # checking to determine whether the error is retryable.
+            text = str(error)
+            # NOTE: we have also seen this, which may warrant retrying:
+            # `requests.exceptions.ConnectionError: ('Connection aborted.',
+            # RemoteDisconnected('Remote end closed connection without
+            # response'))`
+            if 'NewConnectionError' in text or 'Max retries' in text:
+                return True
+
+        return False
+
+    def get_retry_delay(self, retries, response=None):
+        delay = 0
+
+        # As of 2023-11-27, the Wayback Machine does not set a `Retry-After`
+        # header, so this parsing is really just future-proofing.
+        if response is not None:
+            delay = parse_retry_after(response.headers.get('Retry-After')) or delay
+            if response.status_code == 429 and delay == 0:
+                delay = DEFAULT_RATE_LIMIT_DELAY
+
+        # No default backoff on the first retry.
+        if retries > 0:
+            delay = max(self.backoff * 2 ** (retries - 1), delay)
+
+        return delay
+
+
+class WaybackRequestsAdapter(RetryAndRateLimitAdapter, DisableAfterCloseAdapter, RequestsAdapter):
+    def __init__(
+        self,
+        retries: int = 6,
+        backoff: Real = 2,
+        timeout: Union[Real, Tuple[Real, Real]] = 60,
+        user_agent: str = None,
+        memento_rate_limit: Union[RateLimit, Real] = DEFAULT_MEMENTO_RATE_LIMIT,
+        search_rate_limit: Union[RateLimit, Real] = DEFAULT_CDX_RATE_LIMIT,
+        timemap_rate_limit: Union[RateLimit, Real] = DEFAULT_TIMEMAP_RATE_LIMIT,
+    ) -> None:
+        super().__init__(
+            retries=retries,
+            backoff=backoff,
+            rate_limits={
+                '/web/timemap': RateLimit.make_limit(timemap_rate_limit),
+                '/cdx': RateLimit.make_limit(search_rate_limit),
+                # The memento limit is actually a generic Wayback limit.
+                '/': RateLimit.make_limit(memento_rate_limit),
+            }
+        )
+        self.timeout = timeout
+        self.headers = {
+            'User-Agent': (user_agent or
+                           f'wayback/{__version__} (+https://github.com/edgi-govdata-archiving/wayback)'),
+            'Accept-Encoding': 'gzip, deflate'
+        }
```

Comment:
Now that I’m restructuring things, I wonder a bit whether the defaults for `user_agent` and the rate limits belong here, or in `WaybackClient`. As it is, someone wanting to provide a custom adapter needs to make sure they set the right defaults, which are very much tied to “correct” functioning of the overall library.

OTOH, keeping them here keeps configuration simple — you don’t really provide options to `WaybackClient` other than an instance of `WaybackHttpAdapter`, and all the various config options live on the adapter.